pulsar-client/src/main/java/org/apache/pulsar/client/impl/MessageImpl.java

```diff
@@ -489,9 +490,9 @@ private T decodeBySchema(byte[] schemaVersion) {
             return value;
         }
         if (null == schemaVersion) {
-            return schema.decode(getData());
+            return schema.decode(new ByteBufInputStream(payload));
```

Comment:
I presume this bit is the optimisation that prevents the payload copy?

Reply:
yes, wrap ByteBuf to InputStream, and SchemaReader supports decode InputStream, it won't read all bytes into a byte array directly.